dmutils/status.py

```diff
@@ -47,3 +48,68 @@ def enabled_since(date_string):
+    """A stub class to use when implementing additional checks for an app's _status endpoint. See API for example.
+    When raising a StatusError, make sure that the message passed in uniquely identifies the additional check you are
+    performing so that any errors can more easily be tied back to a specific dependency that has failed."""
+    message = ''
+
+    def __init__(self, message):
+        super().__init__(message)
+        self.message = message
+
+
+def get_app_status(data_api_client=None,
+                   search_api_client=None,
+                   ignore_dependencies=False,
+                   additional_checks=None):
+    """Generates output for `_status` endpoints on apps
+
+    :param current_app: The flask `current_app` global.
+    :param data_api_client: The app's data_api_client, if used.
+    :param search_api_client: The app's search_api_client, if used.
+    :param ignore_dependencies: Minimal endpoint checks only (i.e. the app is routable and disk space is fine).
+    :param additional_checks: A sequence of callables that return dicts of data to be injected into the final JSON
+                              response or raise StatusErrors if they need to log an error that should fail the
+                              check (this will cause the endpoint to return a 500).
+    :return: A dictionary describing the current state of the app with, at least, a 'status' key with a value of 'ok'
+             or 'error'.
+    """
+    error_messages = []
+    response_data = {'status': 'ok'}
+
+    disk_status = get_disk_space_status()
+    response_data['disk'] = f'{disk_status[0]} ({disk_status[1]}% free)'
+    if disk_status[0].lower() != 'ok':
+        error_messages.append(f'Disk space low: {disk_status[1]}% remaining.')
+
+    if not ignore_dependencies:
+        response_data['version'] = current_app.config['VERSION']
+        response_data['flags'] = get_flags(current_app)
+
+        if data_api_client:
+            data_api_status = data_api_client.get_status() or {'status': 'n/a'}
+            response_data['api_status'] = data_api_status
+            if data_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Data API.')
+
+        if search_api_client:
+            search_api_status = search_api_client.get_status() or {'status': 'n/a'}
+            response_data['search_api_status'] = search_api_status
+            if search_api_status['status'].lower() != 'ok':
+                error_messages.append('Error connecting to the Search API.')
+
+        for additional_check in (additional_checks or []):
+            try:
+                response_data.update(additional_check())
+
+            except StatusError as e:
+                error_messages.append(e.message)
+
+    if error_messages:
+        response_data['status'] = 'error'
+        response_data['message'] = error_messages
+
+    return jsonify(**response_data), 200 if not error_messages else 500
```

Comment:
I realize it's too late but I *much* preferred the `**{**{...}, ...}` method. Much clearer than having side-effect-y branchy code all up the wang up there. Sure, if you don't know python fully, you don't know what it means the first time you see it. But then you look it up. And then you know *totally unambiguously* what pattern it implements and how it's being used there.

Reply:
It's not too late. 🤔

Reply:
Do we have to accept that every PR will now be a battle ground between ppl that prefer functional style over OO. Woop, Python.

Reply:
The other way is no less oo than otherwise. It's imperative vs functional.

Reply:
Egh I'm not sure either way. I'm going to leave it as it is as I've removed the extra dict now, so there's nothing to unpack together. I think the extra dict was probably an unneeded bit of fluff here so I'm not sure there's value adding it back.

Reply:
If you don't allow mutable objects, then it's barely OO in my opinion ;-)

Reply:
It's not disallowing mutation, just not encouraging its use where it's not necessary, like doing something that should be straightforward like assembling a dictionary. Mutation should be used where it is part & parcel of what the app is trying to _do_.

Reply:
Late to the party here, but what was the `**{**{...}, ...}` method?

Reply:
```
def echo(**kwargs):
  pass

a = {'a': 1}
a2 = {'a': 2}
b = {'b': 1}

{**a, **b} == {'a': 1, 'b': 1}
{**a, **a2} == {'a': 2}
{**a2, **a} == {'a': 1}

echo(**{**a, **a2}) == echo(a=2)
```

Reply:
@Wynndow it's basically a side-effect-free, in-place dictionary merge.